### PR TITLE
Add proposed "Implementers' Checklists" Appendix

### PR DESCRIPTION
**Pull request by cabo**

### CODE REVIEW

File: draft-ietf-cbor-cde.md [hunk 1]
```diff
@@ -302,6 +307,149 @@ This document requests IANA to register the contents of
...
+
+### Preferred serialization encoders
+
+1. Shortest-form encoding of the argument MUST be used for all major
+   types.
+   Major type 7 is used for floating-point and simple values; floating
+   point values have its specific rules for how the shortest form is
+   derived for the argument.
+   The shortest form encoding for any argument that is not a floating
+   point value is:
```
Comment (laurencelundblade) [inline draft-ietf-cbor-cde.md:372]: Type 7 described in 8949 is deterministic without further specification. In particular, the text for simple values requires shortest-form. That's the reason I said all but type 7 and didn't call out an exemption for floats.

But, yes in the obvious implementation, the exception is floats.

I'm fine either way with the text.

File: draft-ietf-cbor-cde.md [hunk 2]
```diff
@@ -302,6 +307,149 @@ This document requests IANA to register the contents of
...
 
+# Implementers' Checklists {#impcheck}
+
+This appendix is informative.
+It provides brief checklists that implementers can use to check their
+implementations.
+It uses {{RFC2119}} language, specifically the keyword MUST, to highlight
+the specific items that implementers may want to check.
+It does not contain any normative mandates.
+This appendix is informative.
```
Comment (laurencelundblade) [inline draft-ietf-cbor-cde.md:318]: Duplicate sentence.

File: draft-ietf-cbor-cde.md [hunk 3]
```diff
@@ -302,6 +307,149 @@ This document requests IANA to register the contents of
...
+
+Notes:
+
+* This is largely a restatement of parts of {{Section 4 of
+  RFC8949@-cbor}}.
+  The purpose of the restatement is to aid the work of implementers,
+  not to redefine anything.
+
+* Duplicate map keys are never valid in CBOR at all (see, for
+  instance, list item "Major type 5" in {{Section 3.1 of RFC8949@-cbor}})
```
Comment (laurencelundblade) [inline draft-ietf-cbor-cde.md:328]: I would remove "for instance". That text in 3.1 is _the_ normative unequivocal statement that duplicate map keys are not allowed. I kind of wish that text in 8949 was stronger.

Understand that some decoders might not detect this and that we have to allow such decoders, but that doesn't change the hard requirement that duplicate map keys are never allowed, right?

File: draft-ietf-cbor-cde.md [hunk 4]
```diff
@@ -302,6 +307,149 @@ This document requests IANA to register the contents of
...
+  the data model by restricting some values and ranges.
+
+* CBOR decoders in general are not required to check for Preferred
+  Serialization or CDE and reject inputs that do not do not fulfill
+  their requirements..
+  However, in an environment that employs deterministic encoding, this
+  negates many of its benefits.
+  Decoder implementations that advertise "support" for Preferred
+  Serialization or CDE need to check the encoding and reject
+  input that is not encoded to the encoding specification in use.
```
Comment (laurencelundblade) [inline draft-ietf-cbor-cde.md:349]: The text "support" is kind of being used as a form of 2119 SHOULD here. I'd much prefer saying decoders MAY support checking.  I don't think it is a SHOULD and would prefer to stick with 2119 terminology.

File: draft-ietf-cbor-cde.md [hunk 5]
```diff
@@ -302,6 +307,149 @@ This document requests IANA to register the contents of
...
+   strings MUST be accepted.
+
+1. If floating-point numbers are supported, the following apply:
+
+   * Half-precision values MUST be accepted.
+   * Double- and single-precision values SHOULD be accepted; leaving these out
+     is only foreseen for decoders that need to work in exceptionally
+     constrained environments.
+   * If double-precision values are accepted, single-precision values
+     MUST be accepted.
```
Comment (laurencelundblade) [inline draft-ietf-cbor-cde.md:432]: I like this wording. I was trying to figure out something like this myself.

File: draft-ietf-cbor-cde.md [hunk 6]
```diff
@@ -302,6 +307,149 @@ This document requests IANA to register the contents of
...
+  the data model layer.
+  Application profiles such as dCBOR are different as they can affect
+  the data model by restricting some values and ranges.
+
+* CBOR decoders in general are not required to check for Preferred
+  Serialization or CDE and reject inputs that do not do not fulfill
+  their requirements..
+  However, in an environment that employs deterministic encoding, this
+  negates many of its benefits.
+  Decoder implementations that advertise "support" for Preferred
```
Comment (laurencelundblade) [inline draft-ietf-cbor-cde.md:347]: But, it doesn't negate the main benefit of Preferred which is interoperability nor the main benefit of CDE which is determinism.

Having spent a year plus trying to understand benefit and negatives of stuff in COSE and HPKE, I'd much prefer explicit statement of negatives and benefits.

**Review by laurencelundblade — COMMENTED** [review]
Hi Carsten, looks generally good. Thank you kindly for including. I made some comments, but am fine with merging as is.

File: draft-ietf-cbor-cde.md (hunk 2 above)
Comment (cabo) [reply draft-ietf-cbor-cde.md:318]: Yes!  That is the point...  We really mean this.

File: draft-ietf-cbor-cde.md (hunk 3 above)
Comment (cabo) [reply draft-ietf-cbor-cde.md:328]: Correct, at least if you want to have valid encoded data items.
Removed the "for instance", but there are of course other places in 8949 that add information to this.
But citing them all here does not make sense, indeed.

File: draft-ietf-cbor-cde.md (hunk 4 above)
Comment (cabo) [reply draft-ietf-cbor-cde.md:349]: This is really a "truth of advertising" statement, not a SHOULD.
So it is phrased as a factual "need to".

File: draft-ietf-cbor-cde.md (hunk 6 above)
Comment (cabo) [reply draft-ietf-cbor-cde.md:347]: Not checking removes some security properties.

Section 5 (security considerations):

```
However, this effect only accrues if the decoder actually checks that
deterministic encoding was applied correctly.
```

This is just restated here...

**Pull request merged**